x-pack/metricbeat/module/aws/awshealth/awshealth.go

```diff
@@ -0,0 +1,436 @@
+func (m *MetricSet) getEventsSummary(
+	ctx context.Context,
+	awsHealth *health.Client,
+) []mb.Event {
+	var events []mb.Event
+	eventFilter := types.EventFilter{
+		EventStatusCodes: []types.EventStatusCode{
+			types.EventStatusCodeUpcoming,
+			types.EventStatusCodeOpen,
+		},
+	}
+
+	var (
+		nextTokenString string
+		eventOutput     *health.DescribeEventsOutput
+		err             error
+		wg              sync.WaitGroup
+	)
+	errCh := make(chan error, maxResults)
+	c := make(chan HealthDetails, maxResults)
+
+	for {
+		// When invoking the DescribeEvents for the first time, there must not exist any NextToken.
+		// DescribeEvents API call will return the next token if there are more records left for querying
+		// If there exist no further records to fetch, next toke will be empty.
+		if nextTokenString == "" {
+			eventOutput, err = awsHealth.DescribeEvents(ctx,
+				&health.DescribeEventsInput{
+					Filter:     &eventFilter,
+					MaxResults: &maxResults,
+				},
+			)
+		} else {
+			eventOutput, err = awsHealth.DescribeEvents(ctx,
+				&health.DescribeEventsInput{
+					Filter:     &eventFilter,
+					MaxResults: &maxResults,
+					NextToken:  &nextTokenString,
+				},
+			)
+		}
+		if err != nil {
+			err = fmt.Errorf("[AWS Health] DescribeEvents failed with : %w", err)
+			m.Logger().Error(err.Error())
+			return nil
+		}
+		ets := eventOutput.Events
+		select {
+		case <-ctx.Done():
+			// Context cancelled, handle graceful termination
+			close(c)
+			return nil
+		default:
+			// Context not cancelled, proceed with the function
+		}
+
+		for _, et := range ets {
+			m.Logger().Debugf("[AWS Health] [Fetch DescribeEventDetails] Event ARN : %s", getStringValueOrDefault(et.Arn))
+			// Increment the WaitGroup counter
+			wg.Add(1)
+			go func(et types.Event) {
+				defer wg.Done() // Decrement the WaitGroup counter when goroutine exits
+				err := m.getDescribeEventDetails(ctx, awsHealth, et, c)
+				if err != nil {
+					errCh <- err
+				}
+			}(et)
+		}
+
+		for i := 0; i < len(ets); i++ {
+			select {
+			case <-ctx.Done():
+				// Context cancelled, handle graceful termination
+				m.Logger().Debug("Context cancelled. Exiting gracefully.")
+				close(c)
+				return nil
+			case err := <-errCh:
+				// Handle errors received from goroutines
+				m.Logger().Error(err.Error())
+			case healthDetails, ok := <-c:
+				if !ok {
+					return nil
+				}
+				m.Logger().Debugf("[AWS Health] [DescribeEventDetails] Event ARN : %s, Affected Entities (Pending) : %d, Affected Entities (Resolved): %d, Affected Entities (Others) : %d", *healthDetails.event.Arn, healthDetails.affectedEntityPending, healthDetails.affectedEntityResolved, healthDetails.affectedEntityOthers)
+				events = append(events, createEvents(healthDetails))
+			}
+		}
+		wg.Wait()
```

Comment:
As you are going this way where you have a buffered channel with a size of 10. The producer is working and producing events for the channel. Meanwhile, the receiver is looping 10 times to either capture the incoming events or any errors that may occur during the process.

Ideally, you do not need the Waitgroups. If you had opted for the other way that I suggested in the diff patch, then it was necessary as closing the channel could only be done when wg.Wait let's pass to the next instruction.

Can you remove the Waitgroup and try? Also, try building the binary with the `-race` flag i.e., `go build -race .` so that you turn on the race detector. It will help us detect if the synchronization done here is race free or not.

Reply:
Validated with `-race`. No issues found

Reply:
How many events were you collecting? Was it a good amount such that 10 goroutines (each time) were getting spawned?

Reply:
Also, remove the WaitGroup and then try. You shouldn't require WaitGroup if you are doing it this way,

Reply:
This may be needed for any rate limiting to AWS health API endpoints, in future.

Reply:
No `Data race` warnings observed when tested with metricbeat with `-race` flag enabled.